Add `round` and `sumAll` functions to `BigDecimal`

jrudder opened the pull request:
Add `round` and `sumAll` functions to `BigDecimal`

<!--
Before submitting a Pull Request, please ensure you've done the following:

- 📖 Read our Contributing Guide: https://github.com/effect-ts/.github/blob/main/CONTRIBUTING.md
- 📖 Read our Code of Conduct: https://github.com/effect-ts/.github/blob/main/CODE_OF_CONDUCT.md
- 👷‍♀️ Create small PRs. In most cases this will be possible.
- 📝 Use descriptive commit messages.
- ✅ Provide tests for your changes if applicable.
- 📗 If your change requires documentation, please update the relevant documentation.
- 📝 Create a changeset for your changes. This helps in tracking and communicating the changes effectively.
- ⏳ Please be patient! We will do our best to review your pull request as soon as possible.

NOTE: Pull Requests from forked repositories will need to be reviewed by a team member before any CI builds will run.
-->

## Type

<!--
What type of change is this? Please check all applicable.
-->

- [ ] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

The `round` function supports rounding at an arbitrary scale/position and the following 10 rounding modes:
- `ceil`: round towards positive infinity
- `floor`: round towards negative infinity
- `to-zero`: round towards zero
- `from-zero`: round away from zero
- `half-ceil`: round to the nearest neighbor; if equidistant round towards positive infinity
- `half-floor`: round to the nearest neighbor; if equidistant round towards negative infinity
- `half-to-zero`: round to the nearest neighbor; if equidistant round towards zero
- `half-from-zero`: round to the nearest neighbor; if equidistant round away from zero
- `half-even`: round to the nearest neighbor; if equidistant round to the neighbor with an even digit at the specified scale
- `half-odd`: round to the nearest neighbor; if equidistant round to the neighbor with an odd digit at the specified scale

The default scale is `0` (round to integer), and the default mode is `half-from-zero`.

The `sumAll` function mirrors that of `BigInt.sumAll`.

## Related

_None_

fubhy commented:
Thanks! Support for rounding has been sorely missing.

fubhy reviewed (changes requested):
Can you make these rounding functions `dual` (data last)

jrudder commented:
@fubhy Thank you for looking into this so quickly.

I have now:

- updated the functions to use `dual`
- cleaned up the JSDoc examples
- changed `digitAt` to use `self` (like the existing `normalize` and other functions, including the others I added)

fubhy commented on packages/effect/src/BigDecimal.ts:
```diff
@@ -1126,3 +1126,234 @@ export const isNegative = (n: BigDecimal): boolean => n.value < bigint0
+ * import { ceil, unsafeFromString } from "effect/BigDecimal"
+ *
+ * assert.deepStrictEqual(ceil(unsafeFromString("145"), -1), unsafeFromString("150"))
+ * assert.deepStrictEqual(ceil(-1)(unsafeFromString("145")), unsafeFromString("150"))
```
Took the liberty to take these out.

fubhy commented on packages/effect/src/BigDecimal.ts:
```diff
@@ -1126,3 +1126,234 @@ export const isNegative = (n: BigDecimal): boolean => n.value < bigint0
+ *      unsafeFromString("2"),
+ *      unsafeFromString("3"),
+ *      unsafeFromString("4")
+ *    ]), unsafeFromString("9"))
```
Took the liberty to adjust this.

jrudder commented on packages/effect/src/BigDecimal.ts:
```diff
@@ -1126,3 +1126,234 @@ export const isNegative = (n: BigDecimal): boolean => n.value < bigint0
+/**
+ * Rounding modes for `BigDecimal`.
+ *
+ * `ceil`: round towards positive infinity
+ * `floor`: round towards negative infinity
+ * `to-zero`: round towards zero
+ * `from-zero`: round away from zero
+ * `half-ceil`: round to the nearest neighbor; if equidistant round towards positive infinity
+ * `half-floor`: round to the nearest neighbor; if equidistant round towards negative infinity
+ * `half-to-zero`: round to the nearest neighbor; if equidistant round towards zero
+ * `half-from-zero`: round to the nearest neighbor; if equidistant round away from zero
+ * `half-even`: round to the nearest neighbor; if equidistant round to the neighbor with an even digit
+ * `half-odd`: round to the nearest neighbor; if equidistant round to the neighbor with an odd digit
```
I wanted to avoid ambiguity of "up", which can mean "toward infinity" (as in Rust) or "away from zero" (as in Java and decimal.js). Similar for "down".

**References:**

[Rust round module](https://docs.rs/libmath/latest/math/round/index.html):
- `ceil`: "Round up"
- `floor`: "Round down"
- `half_away_from_zero`: "Round half away from zero"
- `half_down`: "Round half down"
- `half_to_even`: "Round half to nearest even number"
- `half_to_odd`: "Round half to nearest odd number"
- `half_towards_zero`: "Round half towards zero"
- `half_up`: "Round half up"
- `stochastic`: "Round half randomly up or down"

[Java RoundingMode](https://docs.oracle.com/en/java/javase/24/docs/api/java.base/java/math/RoundingMode.html):
- `CEILING`: "Rounding mode to round towards positive infinity"
- `DOWN`: "Rounding mode to round towards zero"
- `FLOOR`: "Rounding mode to round towards negative infinity"
- `HALF_DOWN`: "...unless both neighbors are equidistant, in which case round down."
- `HALF_EVEN`: "...in which case, round towards the even neighbor"
- `HALF_UP`: "...in which case round up"
- `UNNECESSARY`: "...no rounding is necessary"
- `UP`: "Rounding mode to round away from zero"

[decimal.js rounding modes](https://mikemcl.github.io/decimal.js/#modes):
- `ROUND_UP`: "Rounds away from zero"
- `ROUND_DOWN`: "Rounds towards zero"
- `ROUND_CEIL`: "Rounds towards Infinity"
- `ROUND_FLOOR`: "Rounds towards -Infinity"
- `ROUND_HALF_UP`: "...away from zero"
- `ROUND_HALF_DOWN`: "...towards zero"
- `ROUND_HALF_EVEN`: "...towards even neighbour"
- `ROUND_HALF_CEIL`: "...towards Infinity"
- `ROUND_HALF_FLOOR`: "...towards -Infinity"
- `EUCLID`: "Not a rounding mode"

fubhy replied:
Ok, that makes sense. Ty.

jrudder commented on packages/effect/test/BigDecimal.test.ts:
```diff
@@ -361,6 +361,181 @@ describe("BigDecimal", () => {
+          `Expected ${value} to truncate ${toZero} at scale ${scale}`
+        )
+        assertEquals(
+          BigDecimal.round(value, { mode: "from-zero", scale }),
+          fromZero,
+          `Expected ${value} to round to ${fromZero} with mode fromZero, scale ${scale}`
+        )
+        assertEquals(
+          BigDecimal.round(value, { mode: "half-even", scale }),
+          halfEven,
+          `Expected ${value} to round to ${halfEven} with mode halfEven, scale ${scale}`
+        )
+        assertEquals(
+          BigDecimal.round(value, { mode: "half-odd", scale }),
+          halfOdd,
+          `Expected ${value} to round to ${halfOdd} with mode halfOdd, scale ${scale}`
+        )
+        assertEquals(
+          BigDecimal.round(value, { mode: "half-ceil", scale }),
+          halfCeil,
+          `Expected ${value} to round to ${halfCeil} with mode halfCeil, scale ${scale}`
+        )
+        assertEquals(
+          BigDecimal.round(value, { mode: "half-floor", scale }),
+          halfFloor,
+          `Expected ${value} to round to ${halfFloor} with mode halfFloor, scale ${scale}`
+        )
+        assertEquals(
+          BigDecimal.round(value, { mode: "half-to-zero", scale }),
+          halfToZero,
+          `Expected ${value} to round to ${halfToZero} with mode halfToZero, scale ${scale}`
+        )
+        assertEquals(
+          BigDecimal.round(value, { mode: "half-from-zero", scale }),
+          halfFromZero,
+          `Expected ${value} to round to ${halfFromZero} with mode halfFromZero, scale ${scale}`
+        )
+      }
+    }
+  })
```
Absolutely. I pulled this from the [Unicode ICU rounding modes page](https://htmlpreview.github.io/?https://github.com/unicode-org/icu-demos/blob/main/roundmode/round.html) to quickly get a test-set going, but I can certainly swap it out with hand-picked values that focus on transition spots and edge-cases.

jrudder replied:
I've swapped out the table tests in commit [b4d1d24](https://github.com/Effect-TS/effect/pull/4920/commits/b4d1d2442eb2bdd366c70b997eacb6310a3789fa)
